packages/core/src/browser/label-provider.ts

```diff
@@ -48,6 +49,17 @@ export interface LabelProviderContribution {
      */
     getLongName?(element: object): string;
 
+    /**
+     * Emit when something has changed that may result in this label provider returning a different
+     * value for one or more properties (name, icon etc).
+     */
+    readonly onDidChange?: Event<DidChangeLabelEvent>;
+
+    readonly getConstituentUris?: (compositeUri: URI) => URI[];
```

Comment:
I thought we agree that `URI` should not leak here. Could we remove it and use only `onDidChange`? Implementation should check whether passed element in `affects` actually is `URI`. We would like to pass other objects like `FileStat` and any other shapes having `uri` property.

Reply:
Unfortuntely one cannot implement this in onDidChange using the obvious way of putting something like the following in the DiffUriLabelProviderContribution constructor:
`
labelProvider.onDidChange(event => {

     this.onDidChangeEmitter.fire({
         affects: (element: URI) => {
             if (!(element instanceof URI && DiffUris.isDiffUri(element))) {
                 return false;
             }
             const label = element.path.toString();
             if (label) {
                 return false;
             }
             const [left, right] = DiffUris.decode(element);
             return event.affects(left) || event.affects(right);
         }
     });
});
`
The problem is that the code that fires the event does not know what URIs it may get.  It could potentially get a diff URI in which one of the nested URIs is also a diff URI, down any number of levels.  This above code will recurse indefinitely as it fires an event for a one-level nested diff URI, then an event for a two level nested diff URI etc.  There are ways of using flags to stop this, given that we know consumers will never nest diff URIs.  However this felt to me more like a hack than a generic solution.  So I came up with the current implementation which I feel is more robust in the sense that it is harder for contribution providers to make it all go horribly wrong.